### PR TITLE
stage2: improvements to LLVM backend

### LINKED ISSUE #7688: avoid using the global LLVM context
For example:

https://github.com/ziglang/zig/blob/66e5e92a3e958c8097a2366ff74f34ebdad40154/src/llvm_bindings.zig#L114

This function should never be called and should be deleted from the bindings. Instead we should use `LLVMIntTypeInContext`. Here's the relevant .h snippet:

```c
/**
 * Obtain an integer type from a context with specified bit width.
 */
LLVMTypeRef LLVMInt1TypeInContext(LLVMContextRef C);
LLVMTypeRef LLVMInt8TypeInContext(LLVMContextRef C);
LLVMTypeRef LLVMInt16TypeInContext(LLVMContextRef C);
LLVMTypeRef LLVMInt32TypeInContext(LLVMContextRef C);
LLVMTypeRef LLVMInt64TypeInContext(LLVMContextRef C);
LLVMTypeRef LLVMInt128TypeInContext(LLVMContextRef C);
LLVMTypeRef LLVMIntTypeInContext(LLVMContextRef C, unsigned NumBits);

/**
 * Obtain an integer type from the global context with a specified bit
 * width.
 */
LLVMTypeRef LLVMInt1Type(void);
LLVMTypeRef LLVMInt8Type(void);
LLVMTypeRef LLVMInt16Type(void);
LLVMTypeRef LLVMInt32Type(void);
LLVMTypeRef LLVMInt64Type(void);
LLVMTypeRef LLVMInt128Type(void);
LLVMTypeRef LLVMIntType(unsigned NumBits);
```

There are many more instances such as these.

This accomplishes 2 things:
 1. makes the zig compiler as a library play nicely with other code
 2. paves the way for multi-threaded stuff

For an example of (2) we could build compiler-rt and the main application simultaneously. But if we used the same `LLVMContext` for both, it would cause data races.

The `LLVMContext` can be created and destroyed like this:

```c
/**
 * Create a new context.
 *
 * Every call to this function should be paired with a call to
 * LLVMContextDispose() or the context will leak memory.
 */
LLVMContextRef LLVMContextCreate(void);

/**
 * Destroy a context instance.
 *
 * This should be called for every call to LLVMContextCreate() or memory
 * will be leaked.
 */
void LLVMContextDispose(LLVMContextRef C);

```

### LINKED ISSUE #7689: put all the LLVM alloca instructions together at the beginning of the function
```zig
fn foo() void {}

export fn henlo() void {
    var x: i32 = 1;
    foo();
    var y: i32 = 1;
}
```

actual output:

```llvm
; ModuleID = 'test'
source_filename = "test"

define void @henlo() {
Entry:
  %0 = alloca i32, align 4
  store i32 1, i32* %0, align 4
  call void @foo()
  %1 = alloca i32, align 4
  store i32 1, i32* %1, align 4
  ret void
}

define void @foo() {
Entry:
  ret void
}
```

expected output:

```llvm
; ModuleID = 'test'
source_filename = "test"

define void @henlo() {
Entry:
  %0 = alloca i32, align 4
  %1 = alloca i32, align 4
  store i32 1, i32* %0, align 4
  call void @foo()
  store i32 1, i32* %1, align 4
  ret void
}

define void @foo() {
Entry:
  ret void
}
```

This ensures that LLVM will emit machine code to adjust the stack pointer all at once at the beginning of the function, rather than multiple times. It makes a big difference for optimizations.

### PR DESCRIPTION
**Pull request by FireFox317**
TLDR: Hello world compiles!

```
extern fn puts(s: [*:0]const u8) c_int;
export fn main() c_int {
    _ = puts("hello world!");
    return 0;
}
```

- Add support for globals
- Fixes #7688 
- Fixed #7689 

See commits for details!

### CODE REVIEW

**Review by andrewrk — APPROVED** [review]
:rocket:

File: test/stage2/llvm_backend.zig
Comment (FireFox317) [reply test/stage2/llvm_backend.zig:32]: MachO/MacOS support has not been hooked up yet, this will happen in a follow up patch :)

**Comment by FireFox317** [comment]
Only change that has been made after the approval is to use `std.cstr.line_sep` in the test.

**Pull request merged**
